### PR TITLE
Implemented Relaxed Distributions

### PR DESCRIPTION
**Pull request by rachtsingh**
Not sure if this is the right name for this distribution (Concrete / GumbelSoftmax are other ideas), but this is what Tensorflow calls it. This PR uses the transforms machinery :)

I had to edit `TransformedDistribution`'s `log_prob` method to take into account `event_shape` - this is probably not the right way to do it, but a quick first try that makes it work. 

- [x] Implement RelaxedOneHotCategorical 
- [x]  Implement RelaxedBernoulli

cc @fritzo

### CODE REVIEW

File: torch/distributions/relaxed_categorical.py
Comment (fritzo) [inline torch/distributions/relaxed_categorical.py]: Could you update the docs and example sample?

File: torch/distributions/relaxed_categorical.py
Comment (fritzo) [inline torch/distributions/relaxed_categorical.py]: `RelaxedCategorical` seems ok too, since `OneHot` is kind of the only relay to relax.

File: torch/distributions/relaxed_categorical.py
Comment (fritzo) [inline torch/distributions/relaxed_categorical.py]: 1. Since `Tensor` and `Variable` are being merged, this might not be necessary.
2. Prefer `torch.is_tensor()`, since `torch.Tensor` is an alias rather than a base class:
```py
isinstance(torch.FloatTensor([0]), torch.Tensor)
# True
isinstance(torch.DoubleTensor([0]), torch.Tensor)
# False
torch.Tensor is torch.FloatTensor
# True
```

File: torch/distributions/relaxed_categorical.py
Comment (fritzo) [inline torch/distributions/relaxed_categorical.py]: Document `temperature` parameter

File: torch/distributions/relaxed_categorical.py
Comment (fritzo) [inline torch/distributions/relaxed_categorical.py]: Is `__neg__` cheaper or more stable than `.mul(-1)`?

File: torch/distributions/relaxed_categorical.py
Comment (fritzo) [inline torch/distributions/relaxed_categorical.py]: I would expect either the transform to be `ExpTransform().inv` or the base distribution to be `LogRelaxedOneHotCategorical`. Is the latter a reasonable renaming, analogous to `LogNormal`?

File: torch/distributions/transformed_distribution.py
Comment (fritzo) [inline torch/distributions/transformed_distribution.py]: Some transforms should already know their event shape. Also sometimes `log_prob` is just a number here (e.g. for `identity_transform`). I think it would be safer to do one of:
1. add an `event_shape` arg to pointwise transforms
2. implemement an `TensorTransform(base_transform, event_dim=1)` to wrap `ExpTransform`
3. make this hack a bit more robust:
```py
log_prob = ...sum of log_abs_det_jacobian_terms...
base_log_prob = self.base_dist.log_prob(y)  # knows the correct shape
if not isinstance(log_prob, numbers.Number):
    while log_prob.dim() > base_log_prob.dim():
        log_prob = log_prob.sum(-1)
log_prob += base_log_prob
```

**Review by fritzo — COMMENTED** [review]
Nice factorization into 2 classes!

File: torch/distributions/relaxed_categorical.py
Comment (rachtsingh) [reply torch/distributions/relaxed_categorical.py]: Can I call it via `-(-(uniforms.log()).log())`?

File: torch/distributions/relaxed_categorical.py
Comment (rachtsingh) [reply torch/distributions/relaxed_categorical.py]: 1.  I'm hoping it's not necessary soon, but it definitely is right now because `log_softmax` takes `Variables` and raises when called on a `Tensor`. Can we leave the shim until the merge? 
2. Definitely, didn't realize this was the case.

File: torch/distributions/relaxed_categorical.py
Comment (rachtsingh) [reply torch/distributions/relaxed_categorical.py]: Right, I agree with `LogRelaxedOneHotCategorical` as well. However, TF uses the version I have because that's what the paper names it. My thinking is that it's probably more important just to document it and be consistent with the other interfaces, since no user will need to see it.

Comment (fritzo) [reply torch/distributions/relaxed_categorical.py]: Sounds good, let's keep it standard.

File: torch/distributions/relaxed_categorical.py
Comment (fritzo) [reply torch/distributions/relaxed_categorical.py]: Yes, I just assume it's a tiny bit cheaper.

File: torch/distributions/transformed_distribution.py
Comment (rachtsingh) [reply torch/distributions/transformed_distribution.py]: I like solution `1.` the best. I think fixing the hack as you did doesn't work because `log_prob` isn't a number most of the time - for example it can be `[sample_shape x batch_shape x ...]`. Let me rethink this and get back to you.

Comment (fritzo) [reply torch/distributions/transformed_distribution.py]: :+1: We should also make `constraints` more aware of event dim. @tbrx is already doing this for multivariate normal in #52 and we'll need it e.g. for `AffineOperatorTransform` which transforms from `constraints.real_vector` to `constraints.real_vector`.

Comment (fritzo) [reply torch/distributions/transformed_distribution.py]: After thinking about this a bit, I think a 4th approach could be minimally invasive:
- add a static `event_dim` attribute to each `Transform` (only about 4 lines of code diff)
- update the `log_prob` accumulation loop in `TransformedDistribution.log_prob()`

I'll sketch this in a PR so just to discuss

**Comment by rachtsingh** [comment]
Ok, made changes based on comments, and added the `RelaxedBernoulli` distribution. Blocked on https://github.com/pytorch/pytorch/pull/4937 getting merged, which should be soon :)

**Comment by rachtsingh** [comment]
Added tests that:

1. Rounding the RelaxedBernoulli distribution gives the corresponding Bernoulli distribution
2. Taking the argmax of the RelaxedOneHotCategorical gives the corresponding Categorical distribution
3. As the temperature becomes very large, the first consistently gives 0.5 and the latter gives equal values for each index.

**Review by fritzo — APPROVED** [review]
LGTM

File: torch/distributions/relaxed_categorical.py
Comment (rachtsingh) [reply torch/distributions/relaxed_categorical.py]: We can definitely implement this via a `TransformedDistribution`  of `Gumbel + BoltzmannTransform`; I just didn't see the `BoltzmannTransform`. I'll make an issue about implementing the `log_abs_det_jacobian` and then revisit it when that's been solved?

Comment (fritzo) [reply torch/distributions/relaxed_categorical.py]: I think this will be tricky, but let's move discussion to that issue. Feel free to send this PR upstream before the refactoring.

**Comment by royf** [comment]
@rachtsingh @fritzo 
Are there plans to add KL divergence between two RelaxedOneHotCategorical distributions?
It should be possible to implement it based on this work: https://mast.queensu.ca/~communications/Papers/gil-msc11.pdf (see Proposition 2.3.23).
I may be able to contribute if you could please share a pointer to the math behind the current `log_prob` implementation in https://github.com/pytorch/pytorch/blob/355b24438c1d48ab98948c2ad0c426c6b19f029d/torch/distributions/relaxed_categorical.py#L75.

**Pull request closed without merging**
